src/Interfaces/MOI_interface.jl

```diff
@@ -1101,58 +1083,75 @@ end
     MOI.initialize(model.nlp_data.evaluator, [:Grad,:Hess,:Jac])
 
-    n = num_variables(model)
-    m = num_constraints(model)
-    nnz_hess = get_nnz_hess(model)
-    nnz_jac = get_nnz_jac(model)
-
-    x  = Vector{Float64}(undef,n)
-    g  = Vector{Float64}(undef,m)
-    xl = Vector{Float64}(undef,n)
-    xu = Vector{Float64}(undef,n)
-    zl = Vector{Float64}(undef,n)
-    zu = Vector{Float64}(undef,n)
-
-    l = Vector{Float64}(undef,m)
-    gl = Vector{Float64}(undef,m)
-    gu = Vector{Float64}(undef,m)
-
-    set_x!(model,x,xl,xu,zl,zu)
-    set_g!(model,l,gl,gu)
-
-    obj_scale = get_obj_scale(model.sense)
-    obj(x::StrideOneVector{Float64}) = obj_scale*eval_objective(model,x)
-    obj_grad!(f::StrideOneVector{Float64},x::StrideOneVector{Float64}) =
-        (eval_objective_gradient(model,f,x); obj_scale!=1. && (f.*=obj_scale))
-    con!(c::StrideOneVector{Float64},x::StrideOneVector{Float64}) = eval_constraint(model,c,x)
-    con_jac!(jac::StrideOneVector{Float64},
-             x::StrideOneVector{Float64})=eval_constraint_jacobian(model,jac,x)
-    lag_hess!(hess::StrideOneVector{Float64},x::StrideOneVector{Float64},l::StrideOneVector{Float64},
-              sig::Float64) = eval_hessian_lagrangian(model,hess,x,obj_scale*sig,l)
-    hess_sparsity!(I,J)= hessian_lagrangian_structure(model,I,J)
-    jac_sparsity!(I,J) = jacobian_structure(model,I,J)
+    nvar = num_variables(model)
+    ncon = num_constraints(model)
+    nnzh = get_nnz_hess(model)
+    nnzj = get_nnz_jac(model)
+
+    x0  = Vector{Float64}(undef,nvar)
+    xl = Vector{Float64}(undef,nvar)
+    xu = Vector{Float64}(undef,nvar)
+
+    y0 = Vector{Float64}(undef,ncon)
+    gl = Vector{Float64}(undef,ncon)
+    gu = Vector{Float64}(undef,ncon)
+
+    set_x!(model,x0,xl,xu)
+    set_g!(model,y0,gl,gu)
 
     model.option_dict[:jacobian_constant], model.option_dict[:hessian_constant] = is_jac_hess_constant(model)
-    model.option_dict[:dual_initialized] = !iszero(l)
+    model.option_dict[:dual_initialized] = !iszero(y0)
+
+    meta = NLPModelMeta(
+        nvar,
+        x0 = x0,
+        lvar = xl,
+        uvar = xu,
+        ncon = ncon,
+        y0 = y0,
+        lcon = gl,
+        ucon = gu,
+        nnzj = nnzj,
+        nnzh = nnzh,
+        minimize = model.sense == MOI.MIN_SENSE
+    )
 
-    return NonlinearProgram(n,m,nnz_hess,nnz_jac,0.,x,g,l,zl,zu,xl,xu,gl,gu,obj,obj_grad!,con!,con_jac!,
-                            lag_hess!,hess_sparsity!,jac_sparsity!,INITIAL,Dict{Symbol,Any}())
+    return MadNLPModel(meta,model,NLPModelsCounters())
 end
 
 function MOI.optimize!(model::Optimizer)
-    model.nlp = NonlinearProgram(model)
-    model.ips = Solver(model.nlp;option_dict=copy(model.option_dict))
-    optimize!(model.ips)
+    nlp = MadNLPModel(model)
+    ips = Solver(nlp;option_dict=copy(model.option_dict))
+    model.ips = ips
+    optimize!(ips)
+
+    ips.obj_val = ips.obj_val/ips.obj_scale[]
```

Comment:
I would suggest to move  the lines 1150-1153 in a separate function

Reply:
now we have a separate function `unscale!` and call it at the end of `optimize!`.